### ISSUE TITLE
Add a Specific Implementation for RISC-V ISA AI  Extensions 

### ISSUE THREAD

**Issue by antoinecarme** [antoinecarme/ml2cpp#36]
ML2CPP can generate specific C++ code with RISC-V extensions allowing all scikit-learn, pytorch, caret ML models to be deployable  natively on this platform. 

RISC-V extensions have the advantage to be non-proprietary.

This is a place holder for following the recent efforts on designing RISC-V specific extensions for AI/ML.

No public hardware is available yet.

Sifive Intelligence X280 CPU is an interesting start point. https://www.sifive.com/blog/introduction-to-the-sifive-intelligence-x280

**Comment by antoinecarme** [comment]
![image](https://user-images.githubusercontent.com/16481992/188904447-9477c1ba-cb63-4c63-bdc6-95376e548add.png)

**Comment by antoinecarme** [comment]
https://www.sifive.com/document-file/x280-datasheet

![image](https://user-images.githubusercontent.com/16481992/188904611-d340fb20-77a8-4186-a008-5ad6425610e9.png)

**Comment by antoinecarme** [comment]
https://cambrian-ai.com/sifive-is-leading-the-way-for-innovation-on-risc-v/

![image](https://user-images.githubusercontent.com/16481992/188905490-7ee9875d-465a-42bb-b6f4-4fec88b87949.png)

**Comment by antoinecarme** [comment]
First SBCs with Sifive CPU hit the market (No AI Extensions ;)

https://www.kickstarter.com/projects/starfive/visionfive-2

**Comment by antoinecarme** [comment]
https://github.com/riscv/riscv-v-spec

https://github.com/riscv/riscv-v-spec/blob/master/v-spec.adoc

**Comment by antoinecarme** [comment]
![image](https://user-images.githubusercontent.com/16481992/188979993-ad51ab75-ac93-44a3-a3da-52c7a94ef275.png)

**Comment by antoinecarme** [comment]
[Fastest way to do horizontal SSE vector sum (or other reduction)](https://stackoverflow.com/questions/6996764/fastest-way-to-do-horizontal-sse-vector-sum-or-other-reduction)

**Comment by antoinecarme** [comment]
SiFive Tech Talk on Accelerating AI: Past, Present, and Future 
by Krste Asanovic

https://www.youtube.com/watch?v=8n2HLp2gtYs

“V” Vector Extension Proposal 
Krste Asanovic, UC Berkeley & SiFive

https://www.youtube.com/watch?v=K2CGz1pzNXo

**Comment by antoinecarme** [comment]
https://www.sifive.com/blog/10-important-things-to-know-about-sifive-vectors

![image](https://user-images.githubusercontent.com/16481992/189145306-4fbc4abf-2db9-4bc2-b08c-0afd737a9db7.png)

**Comment by antoinecarme** [comment]
RISC-V Workshop Barcelona May 7th 2018

Vector ISA

Presentation by Roger Espasa at Esperanto Technologies on May 7, 2018 at the RISC-V Workshop in Barcelona, hosted by Barcelona Supercomputing Center and Universitat Politecnica de Catalunya in Barcelona, Spain.

https://www.youtube.com/watch?v=S4fxBZD79gc

**Comment by antoinecarme** [comment]
Try using c++ Standard **Data-parallel vector library** when available (Allow all SIMD implementations : MMX, SSE, AVX-512, ARM Neon, IBM, ...)

https://en.cppreference.com/w/cpp/experimental/simd
https://github.com/VcDevel/std-simd
https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=libstdc%2B%2B-v3/include/experimental/simd;hb=HEAD

**Comment by antoinecarme** [comment]
```
antoine@z600:~/dev/python/packages/reader/ml_reader$ vdir /usr/include/c++/11/experimental/bits/
total 892
-rw-r--r-- 1 root root  10800 Sep  5 21:36 fs_dir.h
-rw-r--r-- 1 root root   9037 Sep  5 21:36 fs_fwd.h
-rw-r--r-- 1 root root   9305 Sep  5 21:36 fs_ops.h
-rw-r--r-- 1 root root  37515 Sep  5 21:36 fs_path.h
-rw-r--r-- 1 root root   2227 Sep  5 21:36 lfts_config.h
-rw-r--r-- 1 root root  10420 Sep  5 21:36 net.h
-rw-r--r-- 1 root root  16452 Sep  5 21:36 numeric_traits.h
-rw-r--r-- 1 root root  20433 Sep  5 21:36 shared_ptr.h
-rw-r--r-- 1 root root 105576 Sep  5 21:36 simd_builtin.h
-rw-r--r-- 1 root root  12711 Sep  5 21:36 simd_converter.h
-rw-r--r-- 1 root root   8995 Sep  5 21:36 simd_detail.h
-rw-r--r-- 1 root root  70919 Sep  5 21:36 simd_fixed_size.h
-rw-r--r-- 1 root root 165341 Sep  5 21:36 simd.h
-rw-r--r-- 1 root root  58319 Sep  5 21:36 simd_math.h
-rw-r--r-- 1 root root  15842 Sep  5 21:36 simd_neon.h
-rw-r--r-- 1 root root   4836 Sep  5 21:36 simd_ppc.h
-rw-r--r-- 1 root root  22818 Sep  5 21:36 simd_scalar.h
-rw-r--r-- 1 root root  82458 Sep  5 21:36 simd_x86_conversions.h
-rw-r--r-- 1 root root 193071 Sep  5 21:36 simd_x86.h
-rw-r--r-- 1 root root   6816 Sep  5 21:36 string_view.tcc

```

**Comment by antoinecarme** [comment]
```
antoine@z600:~$ vdir /usr/include/c++/11/experimental/bits/
fs_dir.h                lfts_config.h           simd_builtin.h          simd.h                  
simd_scalar.h           fs_fwd.h                net.h                   simd_converter.h        
simd_math.h             simd_x86_conversions.h  fs_ops.h                numeric_traits.h        
simd_detail.h           simd_neon.h             simd_x86.h              fs_path.h               
shared_ptr.h            simd_fixed_size.h       simd_ppc.h              string_view.tcc   
```

**Comment by antoinecarme** [comment]
SiFive P670 and P470 RISC-V processors feature RISC-V Vector Extensions

https://www.cnx-software.com/2022/11/02/sifive-p670-and-p470-risc-v-processors-add-risc-v-vector-extensions/

**Comment by antoinecarme** [comment]
RISC-V has standardized a series of standard extensions beyond the integer base instructions which can be implemented or omitted as desired depending on the design goals (e.g. energy/area/performance/storage goals). 

https://en.wikichip.org/wiki/risc-v/standard_extensions

**Comment by antoinecarme** [comment]
![image](https://user-images.githubusercontent.com/16481992/204079331-30631c67-69b3-423a-8a81-a8a0cdda980c.png)

**Comment by antoinecarme** [comment]
https://stackoverflow.com/questions/47405717/dot-product-of-vectors-with-simd

![image](https://user-images.githubusercontent.com/16481992/218606593-bb277237-081b-4140-b948-55cea2be03fc.png)

May be ml2cpp already generates code that a smart compiler can vectorize automatically with SIMD instructions. This is also true for all hardware architectures.

ml2cpp only generated source code, is not responsible of the environment this code will be compiled/used in. PERIOD.
